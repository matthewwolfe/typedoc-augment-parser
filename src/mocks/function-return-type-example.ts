import type { JSONOutput } from 'typedoc';

export const project: JSONOutput.ProjectReflection = {
  id: 0,
  name: 'function-return-type-example',
  variant: 'project',
  kind: 1,
  flags: {},
  children: [
    {
      id: 1,
      name: 'doSomethingFunction',
      variant: 'declaration',
      kind: 64,
      flags: {},
      sources: [
        {
          fileName: 'index.ts',
          line: 1,
          character: 9,
        },
      ],
      signatures: [
        {
          id: 2,
          name: 'doSomethingFunction',
          variant: 'signature',
          kind: 4096,
          flags: {},
          sources: [
            {
              fileName: 'index.ts',
              line: 1,
              character: 9,
            },
          ],
          type: {
            type: 'reflection',
            declaration: {
              id: 3,
              name: '__type',
              variant: 'declaration',
              kind: 65536,
              flags: {},
              children: [
                {
                  id: 6,
                  name: 'isError',
                  variant: 'declaration',
                  kind: 1024,
                  flags: {},
                  sources: [
                    {
                      fileName: 'index.ts',
                      line: 5,
                      character: 4,
                    },
                  ],
                  type: {
                    type: 'intrinsic',
                    name: 'boolean',
                  },
                },
                {
                  id: 5,
                  name: 'isFetching',
                  variant: 'declaration',
                  kind: 1024,
                  flags: {},
                  sources: [
                    {
                      fileName: 'index.ts',
                      line: 4,
                      character: 4,
                    },
                  ],
                  type: {
                    type: 'intrinsic',
                    name: 'boolean',
                  },
                  defaultValue: 'true',
                },
                {
                  id: 4,
                  name: 'isLoading',
                  variant: 'declaration',
                  kind: 1024,
                  flags: {},
                  sources: [
                    {
                      fileName: 'index.ts',
                      line: 3,
                      character: 4,
                    },
                  ],
                  type: {
                    type: 'intrinsic',
                    name: 'boolean',
                  },
                  defaultValue: 'true',
                },
                {
                  id: 7,
                  name: 'setData',
                  variant: 'declaration',
                  kind: 1024,
                  flags: {},
                  sources: [
                    {
                      fileName: 'index.ts',
                      line: 6,
                      character: 4,
                    },
                  ],
                  type: {
                    type: 'reflection',
                    declaration: {
                      id: 8,
                      name: '__type',
                      variant: 'declaration',
                      kind: 65536,
                      flags: {},
                      sources: [
                        {
                          fileName: 'index.ts',
                          line: 6,
                          character: 13,
                        },
                      ],
                      signatures: [
                        {
                          id: 9,
                          name: '__type',
                          variant: 'signature',
                          kind: 4096,
                          flags: {},
                          sources: [
                            {
                              fileName: 'index.ts',
                              line: 6,
                              character: 13,
                            },
                          ],
                          type: {
                            type: 'intrinsic',
                            name: 'void',
                          },
                        },
                      ],
                    },
                  },
                  defaultValue: '...',
                },
              ],
              groups: [
                {
                  title: 'Properties',
                  children: [6, 5, 4, 7],
                },
              ],
              sources: [
                {
                  fileName: 'index.ts',
                  line: 2,
                  character: 9,
                },
              ],
            },
          },
        },
      ],
    },
  ],
  groups: [
    {
      title: 'Functions',
      children: [1],
    },
  ],
  packageName: 'function-return-type-example',
  symbolIdMap: {
    '0': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '',
    },
    '1': {
      sourceFileName: 'src/index.ts',
      qualifiedName: 'doSomethingFunction',
    },
    '2': {
      sourceFileName: 'src/index.ts',
      qualifiedName: 'doSomethingFunction',
    },
    '3': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '__object',
    },
    '4': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '__object.isLoading',
    },
    '5': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '__object.isFetching',
    },
    '6': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '__object.isError',
    },
    '7': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '__object.setData',
    },
    '8': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '__function',
    },
    '9': {
      sourceFileName: 'src/index.ts',
      qualifiedName: '__function',
    },
  },
};
