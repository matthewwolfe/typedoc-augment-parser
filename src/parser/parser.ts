import { setGlobalOptions } from '@pkg/config/global';
import { findChildById } from '@pkg/project/findChildById';
import { findChildByName } from '@pkg/project/findChildByName';
import { findSymbolById } from '@pkg/project/findSymbolById';
import { groups } from '@pkg/project/groups';

import type { JSONOutput } from 'typedoc';
import type { GlobalOptions } from '@pkg/config/global';

function augment(
  project: JSONOutput.ProjectReflection,
  options: Partial<GlobalOptions> = {}
) {
  setGlobalOptions(options);

  return {
    findChildById: findChildById(project),
    findChildByName: findChildByName(project),
    findSymbolById: findSymbolById(project),
    groups: () => groups(project),
    packageName: project.packageName,
    packageVersion: project.packageVersion,
  };
}

export const parser = {
  augment,
} as const;
