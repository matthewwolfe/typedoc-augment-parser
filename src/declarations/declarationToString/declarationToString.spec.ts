import { describe, expect, test } from 'vitest';
import { parser } from '@pkg/parser';
import { project } from '@pkg/mocks/function-return-type-example';

import type { FunctionDeclaration } from '@pkg/types/declarations';

describe('declarationToString', () => {
  const parsed = parser.augment(project);

  test('function return type', () => {
    const child = parsed.findChildByName<FunctionDeclaration>(
      'doSomethingFunction'
    );

    expect(child?.typeToString()).toMatchSnapshot();
  });
});
