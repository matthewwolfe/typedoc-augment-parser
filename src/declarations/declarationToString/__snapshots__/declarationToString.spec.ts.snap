// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`declarationToString > function return type 1`] = `
"() => {
  isError: boolean;
  isFetching?: boolean;
  isLoading?: boolean;
  setData?: () => void;
}"
`;
